happy/HappyProcessStart.py

```diff
@@ -261,32 +261,53 @@ def __start_daemon(self):
             cmd = self.stripRunAsRoot(cmd)
             need_internal_sudo = False
 
-        env_vars = ""
+        env_vars_list = []
+        cmd_list_prefix = []
+        need_bash = False
+
+        if "bash -c" in cmd:
+            tmp = cmd.split("bash -c")
+            need_bash = True
+            cmd_list_prefix = tmp[0].split()
+            cmd = tmp[1]
+
         for key, value in self.env.items():
-            env_vars += key + "=" + value + " "
+            tmp = ""
+            try:
+                tmp = "" + key + "=" + value
+                env_vars_list.append(tmp)
+            except:
+                self.logger.error("Failed to serialize environment variable %s" % (key));
 
-        self.logger.debug("HappyProcessStart with env: > %s" % (env_vars))
+        self.logger.debug("HappyProcessStart with env: > %s" % (env_vars_list))
 
         if self.strace:
-            strace = "strace -tt -o " + self.strace_file + " "
-            for var in env_vars.split():
-                strace += ("-E " + var + " ")
-            cmd = strace + cmd
+            cmd_list_prefix = ["strace", "-tt", "-o", self.strace_file] + cmd_list_prefix
+            tmp = []
+            for i in env_vars_list:
+                tmp.append("-E")
+                tmp.append(i)
+            env_vars_list = tmp
+
         elif need_internal_sudo:
-            cmd = env_vars + cmd
-        elif len(env_vars):
-            cmd = "bash -c " + env_vars + cmd
+            pass
+        elif len(env_vars_list):
+            need_bash = True
 
         if need_internal_sudo:
+            tmp = "ls"
```

Comment:
could you explain why do we change to "ls" command here?

Reply:
This is clearly breaking abstractions.  I'm creating a command to be run as superuser (whatever that requires within happy), using a dummy command (`ls`) and then I'm stripping the command itself, and I process the command prefix to parse whatever was necessary to run the command as the requested user into a list.  A better way to do this would have been to add something like `Driver.getRunAsRootPrefixList()` and `Driver.getRunAsUserPrefixList()`

Reply:
On reflection, I decided that it was the right time to add those abstractions.  The PR has been modified to reflect that.